### PR TITLE
Add analyzer to vectorizer

### LINKED ISSUE #1217: featurizer_count_vectors with words and N-Grams?
Hey,

I need to use N-Grams with featurizer_count_vectors , but I think it would improve importantly the classification if you could also use it together with word itself. Such that not trained words like for the tensorflow_embedding have a chance if they have specific ngrams. 

Can you this easily within the pipelin or is just one possible at the moment?

```
pipeline:
- name: "intent_featurizer_count_vectors"
  analyzer: 'char'
  "min_ngram": 2  # int
  "max_ngram": 3  # int
```

Can you put the featurizer as a series, such that first it creates features with words then you put again in the pipeline the featurizer for `char`?

```
pipeline:
- name: "intent_featurizer_count_vectors"
  analyzer: 'char'
  "min_ngram": 2  # int
  "max_ngram": 3  # int
- name: "intent_featurizer_count_vectors"
  analyzer: 'word'

```

EDIT:

I tried the `intent_featurizer_ngrams` after `intent_featurizer_count_vectors` in my pipeline:

```
language: "de_core_news_sm"

pipeline:
- name: "nlp_spacy"
- name: "tokenizer_spacy"
- name: "intent_featurizer_count_vectors"
- name: "intent_featurizer_ngrams"
  # Maximum number of ngrams to use when augmenting
  # feature vectors with character ngrams
  max_number_of_ngrams: 5
- name: "intent_classifier_sklearn"
```
But this gives error:

```
Traceback (most recent call last):
  File "C:\Users\\Anaconda3\envs\rasa12\lib\site-packages\rasa_nlu\model.py", line 340, in parse
    component.process(message, **self.context)
  File "C:\Users\\Anaconda3\envs\rasa12\lib\site-packages\rasa_nlu\featurizers\ngram_featurizer.py", line 88, in process
    updated = self._text_features_with_ngrams(message, self.best_num_ngrams)
  File "C:\Users\\Anaconda3\envs\rasa12\lib\site-packages\rasa_nlu\featurizers\ngram_featurizer.py", line 97, in _text_features_with_ngrams
    return self._combine_with_existing_text_features(message, extras)
  File "C:\Users\\Anaconda3\envs\rasa12\lib\site-packages\rasa_nlu\featurizers\__init__.py", line 18, in _combine_with_existing_text_features
    additional_features))
  File "C:\Users\Anaconda3\envs\rasa12\lib\site-packages\numpy\core\shape_base.py", line 288, in hstack
    return _nx.concatenate(arrs, 1)
ValueError: all the input arrays must have same number of dimensions
```

### LINKED ISSUE #1494: Check whether Subword Semantic Hashing improve performance of tf flu classifier
The idea is from the paper: https://arxiv.org/pdf/1810.07150.pdf

It should work with current implementation, but using character ngrams in `CountVectorizer`

### LINKED ISSUE #1530: Feature Request: character_count_vectorizer_
**Feature Request**:
It would be useful to have a separate `intent_featurizer_count_char_vectors` (or similarly named) Component. Using a separate Component makes it simple to combine them using defaults (mixing defaults across char- and word-level `CountVectorizer` is messy since they're very different). 

**Sample**:
```yml
  - name: "intent_featurizer_count_vectors"
    max_features: 1000
    min_ngram: 1  # int
    max_ngram: 2  # int

  - name: "charcountvectorizer.charcountvectorizer.CharCountVectorsFeaturizer"
    max_features: 1000
    analyzer: "char"
    min_ngram: 1
    max_ngram: 6
```
This config file would create features for the top 1000 *word* uni- and bi-grams, then would supplement that with the top 1000 *character* uni- to 6-grams. I've tested this with my own data and confirmed it does imbue the pipeline with some spelling/typo invariance. Obviously once it's default component we can name it something better than it is above.

The only changes I think are necessary are:
1) base the new component off `CountVectorFeaturizer`
2) add 'analyzer' argument
3) edit new defaults
4) update `.load` to reference itself instead of `CountVectorFeaturizer`
5) add it to the registry

Could address #1217 and #1494

### LINKED ISSUE #1578: Problem with overriding default values in tensorflow intent classification
**Rasa NLU version**: 0.14.0a1

**Operating system** (windows, osx, ...): Linux

**Issue**:
 `intent_classifier_tensorflow_embedding` does not override the default values.

**How to Reproduce**:
```
- name: "intent_classifier_tensorflow_embedding"
  "epochs": 100
```
does not override the default value of epochs (300) with 100.

### PR DESCRIPTION
**Pull request by Ghostvv**
**Proposed changes**:
- add `analyzer` argument to CountVectorsFeaturizer
- fixes #1530
- fixes #1217
- fixes #1494 
- fixes #1578 

**Status (please check what you already did)**:
- [x] made PR ready for code review
- [X] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog

### CODE REVIEW

**Comment by Ghostvv** [comment]
previous PR:  https://github.com/RasaHQ/rasa_nlu/pull/1540

**Comment by Ghostvv** [comment]
blocked by https://github.com/RasaHQ/rasa_nlu/pull/1546

**Comment by amn41** [comment]
@Ghostvv I would argue this is not blocked by https://github.com/RasaHQ/rasa_nlu/pull/1546 because this is useful on its own (I am using it) - wdyt?

File: tests/base/test_config.py [hunk 1]
```diff
@@ -68,3 +69,17 @@ def test_set_attr_on_component(default_config):
 
     assert cfg.for_component("intent_classifier_sklearn") == expected
     assert cfg.for_component("tokenizer_spacy") == {"name": "tokenizer_spacy"}
+
+
+def test_override_defaults():
```
Comment (Ghostvv) [inline tests/base/test_config.py]: should we iterate over all components and check that defaults are overwritten?

File: rasa_nlu/featurizers/count_vectors_featurizer.py [hunk 2]
```diff
@@ -22,10 +22,15 @@
 class CountVectorsFeaturizer(Featurizer):
     """Bag of words featurizer
 
-    Creates bag-of-words representation of intent features
-    using sklearn's `CountVectorizer`.
-    All tokens which consist only of digits (e.g. 123 and 99
-    but not ab12d) will be represented by a single feature."""
+        Creates bag-of-words representation of intent features
```
Comment (tmbo) [inline rasa_nlu/featurizers/count_vectors_featurizer.py]: please double check what the google python doc comments style guide says about this, but I think there should not be any intendation here (i.e. keep it the way it was).

File: rasa_nlu/featurizers/count_vectors_featurizer.py [hunk 3]
```diff
@@ -132,6 +146,19 @@ def __init__(self, component_config=None):
         # handling Out-Of-Vacabulary (OOV) words
         self._load_OOV_params()
 
+        if self.analyzer != 'word':
```
Comment (tmbo) [inline rasa_nlu/featurizers/count_vectors_featurizer.py]: let's put these validations in a separate function

File: sample_configs/embedding_advanced.yml [hunk 4]
```diff
@@ -0,0 +1,6 @@
+language: en
+pipeline:
```
Comment (tmbo) [inline sample_configs/embedding_advanced.yml]: why is this one "advanced" 😆 ?

File: rasa_nlu/classifiers/embedding_intent_classifier.py [hunk 5]
```diff
@@ -136,7 +136,7 @@ def __init__(self,
         self._check_tensorflow()
         super(EmbeddingIntentClassifier, self).__init__(component_config)
 
-        self._load_params()
+        self._load_params(**component_config)
```
Comment (tmbo) [inline rasa_nlu/classifiers/embedding_intent_classifier.py]: I think this should rather be `self.component_config` and we should remove the deep copy & overwrite implementation in `_load_params`. This behaviour is already implemented in the init method of `Component`.

File: sample_configs/embedding_advanced.yml (hunk 4 above)
Comment (Ghostvv) [reply sample_configs/embedding_advanced.yml]: I believe it is the test config,  @amn41 ?

Comment (amn41) [reply sample_configs/embedding_advanced.yml]: oh because it specifies the components one-by-one rather than the predefined pipeline

Comment (Ghostvv) [reply sample_configs/embedding_advanced.yml]: I moved it to data/test folder since it is only used in the test

**Comment by jstriebel** [comment]
It would be great if you could add a prerelease once this is merged, as it is not possible to specify the embedding parameters in a config currently (see https://github.com/RasaHQ/rasa_nlu/pull/1555 for details).

**Pull request merged**